Update for OpenSSL 1.1 compatibility (Apple-friendly version)

voretaq7 opened the pull request:
This is a redo of PR #320 that doesn't break the build for OS X.
(Cleaned-Up version of PR #326 against upstream - no extraneous commits)

* Handle OpenSSL 1.1 opaque structures for EVP digest functions

* Add openssl_compat.c shims file to support pre-1.1 OpenSSL installs
  (ref. https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes)

* Correctly make the above fixes in do_acksum (Apple HFS checksum) code.

Turns out my office Mac was still set up for building software :)

Build checked OK with the OpenSSL shims on OS X 10.13.6 / OpenSSL 1.0.2d (Homebrew) & generated "a" (Apple HFS) transcript entries with fsdiff -C -c sha1.
Should be double-checked on 10.14 before merging but I don't expect any issues.

voretaq7 commented:
Geez I can't win with git today - didn't re-add the Pre-1.1 compatibility shims (and then re-added an old version of them). Should be finalized now.

Rechecked against FreeBSD (11/OpenSSL 1.0.2o, 12/OpenSSL 1.1.1a) OS X 10.13.6 (OpenSSL 1.0.2d), Linux (OpenSSL 0.9.8g ; OpenSSL 1.1.0g) - all look OK.

Could use a check on OS X with OpenSSL 1.1.x and also against LibreSSL on any platform (I don't have a machine running either of those handy).

hfath commented:
A few comments...

_openssl_compat.c_ -- There is no need for the #ifndef RADMIND_OPENSSL_SHIMS guard, since the .c file is not #included anywhere. OTOH, you may want to #include "openssl_compat.h" to get the prototypes, see <https://github.com/NetBSD/pkgsrc/blob/trunk/sysutils/radmind/patches/patch-openssl__compat.c>.

_openssl_compat.h_ -- You will need to #include <openssl/opensslv.h> to get the openssl version, see <https://github.com/NetBSD/pkgsrc/blob/trunk/sysutils/radmind/patches/patch-openssl__compat.h>.

voretaq7 commented:
All good notes hfath :)

I added the header to pull in the prototypes from `openssl_compat.h` (embarrassing that I forgot to do that - protects against prototype mismatches). This also picks up the OpenSSL version, string.h, and openssl/engine.h as needed so those don't have to be included in the .c file anymore.

Related to that, `openssl/opensslv.h` was pulled in as part of the `openssl/engine.h` chain (I had changed the order of includes so it doesn't need to be pulled in separately), but on reflection I like your way better: Less work for the preprocessor when we're using OpenSSL 1.1.  `opensslv.h` is now pulled in before the version check and everything else is after the version guard only if needed.

The `RADMIND_OPENSSL_SHIMS` guard was left in the old patch out of an abundance of caution (I had originally planned to #include the compat code through the header but switched to building it independently to match the style used in the rest of the radmind codebase and make it easier to replace later) - it had no practical effect & was cleaned up in this PR since it's not necessary.

Re-checked & builds OK on FreeBSD (OpenSSL 1.1.1a) & OS X (OpenSSL 1.0.2d/Homebrew).
I spun down my Linux VMs but I'm pretty sure I've exercised all the paths with those two builds.

magnusviri commented:
Either later this week or next week I’m going to respond to all of this and test and merge as much as possible.

James Reynolds
Sr Systems Administrator
School of Biological Sciences
The University of Utah
801-585-3086

On May 9, 2019, at 12:30 PM, Michael Graziano <notifications@github.com<mailto:notifications@github.com>> wrote:


All good notes hfath :)

I added the header to pull in the prototypes from openssl_compat.h (embarrassing that I forgot to do that - protects against prototype mismatches). This also picks up the OpenSSL version, string.h, and openssl/engine.h as needed so those don't have to be included in the .c file anymore.

Related to that, openssl/opensslv.h was pulled in as part of the openssl/engine.h chain (I had changed the order of includes so it doesn't need to be pulled in separately), but on reflection I like your way better: Less work for the preprocessor when we're using OpenSSL 1.1. opensslv.h is now pulled in before the version check and everything else is after the version guard only if needed.

The RADMIND_OPENSSL_SHIMS guard was left in the old patch out of an abundance of caution (I had originally planned to #include the compat code through the header but switched to building it independently to match the style used in the rest of the radmind codebase and make it easier to replace later) - it had no practical effect & was cleaned up in this PR since it's not necessary.

Re-checked & builds OK on FreeBSD (OpenSSL 1.1.1a) & OS X (OpenSSL 1.0.2d/Homebrew).
I spun down my Linux VMs but I'm pretty sure I've exercised all the paths with those two builds.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/Radmind/radmind/pull/327#issuecomment-491016147>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAFNUZJMEXUZVED55BP6K6LPURUWDANCNFSM4HLV4WLQ>.
